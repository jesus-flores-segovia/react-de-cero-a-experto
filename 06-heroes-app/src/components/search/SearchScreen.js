import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { useLocation } from 'react-router';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({history}) => {

    const location = useLocation();
    const q = queryString.parse(location.search).q;

    const [values, handleInputChange] = useForm({
        query: q || ""
    });

    const {query} = values;

    let filteredHeroes = useMemo(() => {
        return getHeroesByName(q);
    }, [q]);

    const handleSearch = (e) => {
        e.preventDefault();

        if (!query) {
            history.replace({ pathname: location.pathname, search: ''});
            return;
        }
             
        location.search = queryString.stringify({q: query});
        history.push(`?${location.search}`);    
    }

    return (
        <div>
            <h1>SearchScreen</h1>
            <hr/>

            <div className="row">

                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr/>

                    <form onSubmit={handleSearch}>
                        <input 
                            className="form-control"
                            name="query"
                            type="text"
                            id="query"
                            value={query}
                            placeholder="Find your hero"
                            autoComplete="off"
                            onChange={handleInputChange}
                        />
                        <button 
                            type="submit"
                            className="btn btn-block btn-outline-primary w-100 mt-2"
                        >
                            Search
                        </button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>
                    {
                        filteredHeroes.length === 0 && 
                        !q &&
                            <div className="alert alert-info" role="alert">
                                <p>
                                    Try to search something.
                                </p>
                            </div>
                    }
                    {
                        filteredHeroes.length === 0 && 
                        q &&
                            <div className="alert alert-danger" role="alert">
                                <p>
                                    There are not results for '{query}'. <br/>
                                    Please try again.
                                </p>
                            </div>
                    }
                    <div className="row">
                        {
                            filteredHeroes.length !== 0 && filteredHeroes.map(hero => 
                                <HeroCard 
                                    key={hero.id}
                                    {...hero}
                                />
                            )
                        }
                    </div>
                </div>
            </div>

        </div>
    )
}
