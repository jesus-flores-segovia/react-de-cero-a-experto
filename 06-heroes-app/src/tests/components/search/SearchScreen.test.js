import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { SearchScreen } from "../../../components/search/SearchScreen"

describe("Tests inside file 'SearchScreen.test.js'", () => {

    test("SearchScreen component must be shown correctly", () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find("h1").text()).toBe("SearchScreen");
    })

    test("Must show a hero result when a query param is setted", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find("h5").text()).toBe("Batman");
        expect(wrapper.find("input").prop("value")).toBe("batman");
    })

    test("Must show an alert that advice the user that there are not results for his search", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman1"]}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert.alert-danger p").text().trim()).toBe("There are not results for 'batman1'. Please try again.")
    })

    test("Must calls the push function from history and rewrite the url with the new query param", () => {
        const history = {
            push: jest.fn()
        };
        
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route path="/search" component={() => <SearchScreen history={history}/>}/>
            </MemoryRouter>
        );

        wrapper.find("input").simulate("change", {
            target: {
                name: "query",
                value: "batman"
            }
        });

        wrapper.find("form").prop("onSubmit")({
            preventDefault(){}
        });

        expect(history.push).toHaveBeenCalledWith("?q=batman");
    })
})