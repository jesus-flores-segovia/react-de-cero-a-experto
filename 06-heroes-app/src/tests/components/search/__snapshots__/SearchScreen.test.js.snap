// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file 'SearchScreen.test.js' Must show a hero result when a query param is setted 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          id="query"
          name="query"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-block btn-outline-primary w-100 mt-2"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="row"
      >
        <div
          className="col-xl-3 col-lg-3 col-md-4 col-sm-6 mb-4"
        >
          <div
            className="card h-100"
          >
            <img
              alt="Batman"
              className="card-img-top"
              src="./assets/heroes/dc-batman.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                See more
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests inside file 'SearchScreen.test.js' Must show an alert that advice the user that there are not results for his search 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          id="query"
          name="query"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman1"
        />
        <button
          className="btn btn-block btn-outline-primary w-100 mt-2"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
        role="alert"
      >
        <p>
          There are not results for '
          batman1
          '. 
          <br />
          Please try again.
        </p>
      </div>
      <div
        className="row"
      />
    </div>
  </div>
</div>
`;

exports[`Tests inside file 'SearchScreen.test.js' SearchScreen component must be shown correctly 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          id="query"
          name="query"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-block btn-outline-primary w-100 mt-2"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
        role="alert"
      >
        <p>
          Try to search something.
        </p>
      </div>
      <div
        className="row"
      />
    </div>
  </div>
</div>
`;
